### PR TITLE
CF-296 Refactor test_verify_dst_functionality.py

### PR DESCRIPTION
**Pull request by plomakin**
- Moved duplicated checks to separate methods
- Added floating IP assignment check and retry
- Added vm.name logging when fail occurs

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.tests.base
devlab/tests/base.py:286: [W0102(dangerous-default-value), 
BasePrerequisites.wait_until_objects_created] Dangerous default value [] as 
argument

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: devlab/tests/base.py [hunk 1]
```diff
@@ -274,14 +274,29 @@ def check_snapshot_state(self, snp_id):
...
+        for net in vm.addresses:
+            for addr in vm.addresses[net]:
+                if addr['OS-EXT-IPS:type'] == 'floating':
+                    return True
+        vm.add_floating_ip(floating_ip_address)
+        return False
+
+    @staticmethod
+    def wait_until_objects_created(obj_list, check_func, timeout,
+                                   param_list=[]):
```
Comment (roman-verchikov) [inline devlab/tests/base.py]:
```
devlab/tests/base.py:286: [W0102(dangerous-default-value), 
BasePrerequisites.wait_until_objects_created] Dangerous default value [] as 
argument
```

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: devlab/tests/base.py [hunk 2]
```diff
@@ -274,14 +274,28 @@ def check_snapshot_state(self, snp_id):
...
     @staticmethod
+    def check_floating_ip_assigned(vm, floating_ip_address):
+        for net in vm.addresses:
+            for addr in vm.addresses[net]:
+                if addr['OS-EXT-IPS:type'] == 'floating':
+                    return True
+        try:
+            vm.add_floating_ip(floating_ip_address)
+        except ks_exceptions.BadRequest:
+            pass
```
Comment (roman-verchikov) [inline devlab/tests/base.py]: 1. Let's catch all nova client exceptions;
2. Let's log that event

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: devlab/tests/base.py [hunk 3]
```diff
@@ -274,14 +274,28 @@ def check_snapshot_state(self, snp_id):
...
         waiting = 0
         delay = 1
         while waiting < timeout:
             for obj in obj_list[:]:
-                if check_func(obj):
-                    obj_list.remove(obj)
+                if not isinstance(obj, tuple):
+                    obj = (obj, )
+                if check_func(*obj):
+                    obj_list.remove(obj[0])
```
Comment (roman-verchikov) [inline devlab/tests/base.py]: Looks like the `remove` will not remove object from the list if it's a tuple of more than 1 element

File: devlab/tests/base.py [hunk 4]
```diff
@@ -274,14 +274,28 @@ def check_snapshot_state(self, snp_id):
...
         obj_list = obj_list[:]
         waiting = 0
         delay = 1
         while waiting < timeout:
             for obj in obj_list[:]:
-                if check_func(obj):
-                    obj_list.remove(obj)
+                if not isinstance(obj, tuple):
+                    obj = (obj, )
+                if check_func(*obj):
```
Comment (SVilgelm) [inline devlab/tests/base.py]:
```python
for obj in obj_list[:]:
    if check_func(*obj if isinstance(obj, tuple) else obj):
        ...
```

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: devlab/tests/base.py (hunk 4 above)
Comment (antonf) [inline devlab/tests/base.py]: Nope, it won't work

    >>> def foo(*x): return x
    ... 
    >>> foo(*(1,2) if False else 3)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: foo() argument after * must be a sequence, not int

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.tests.base
devlab/tests/base.py:254: [W0211(bad-staticmethod-argument), 
BasePrerequisites.check_vm_state] Static method with 'self' as first argument

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/2802/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.tests.base
devlab/tests/base.py:258: [E0102(function-redefined), 
BasePrerequisites.check_vm_state] method already defined line 253

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
:+1:

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by plomakin** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: devlab/tests/base.py [hunk 5]
```diff
@@ -274,14 +277,30 @@ def check_snapshot_state(self, snp_id):
...
+    @staticmethod
     def wait_until_objects_created(obj_list, check_func, timeout):
         obj_list = obj_list[:]
         waiting = 0
         delay = 1
         while waiting < timeout:
             for obj in obj_list[:]:
-                if check_func(obj):
-                    obj_list.remove(obj)
+                if isinstance(obj, tuple):
```
Comment (SVilgelm) [inline devlab/tests/base.py]:
```python
if not isinstance(ob, tuple):
    obj = (obj, )
if check_func(*obj):
    obj_list.remove(obj)
```
???

File: devlab/tests/testcases/test_verify_dst_functionality.py [hunk 6]
```diff
@@ -23,10 +23,12 @@
 
 import devlab.tests.config as config
 import devlab.tests.functional_test as functional_test
+from devlab.tests.base import BasePrerequisites as base
```
Comment (SVilgelm) [inline devlab/tests/testcases/test_verify_dst_functionality.py]: If it possible, let's start using the general openstack rules: http://docs.openstack.org/developer/hacking/#imports

File: devlab/tests/base.py (hunk 5 above)
Comment (SVilgelm) [inline devlab/tests/base.py]: Let's save the lines of code when it possible and doesn't spoil the readable.

Comment (plomakin) [inline devlab/tests/base.py]: In this case we can't just remove tuple from obj_list, if originally it was not a list of tuples

Comment (SVilgelm) [inline devlab/tests/base.py]: sorry, i forgot
what about this:
```python
args = obj if isinstance(obj, tuple) else (obj, )
if check_func(*argsj):
    obj_list.remove(obj)
```

Comment (plomakin) [inline devlab/tests/base.py]: k. thanks

File: devlab/tests/testcases/test_verify_dst_functionality.py (hunk 6 above)
Comment (plomakin) [inline devlab/tests/testcases/test_verify_dst_functionality.py]: I'm using Pycharm "organize imports", not sure about openstack's one. Are we using it in CF project?

Comment (plomakin) [inline devlab/tests/testcases/test_verify_dst_functionality.py]: Ok, i think we'll move this comment to a different task. I will organize all tests imports when will do a package for it.

Comment (SVilgelm) [inline devlab/tests/testcases/test_verify_dst_functionality.py]: No, you don't need to reorganize all import, just do it for new or current modules if you modify imports.

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
can be merged

**Pull request merged**